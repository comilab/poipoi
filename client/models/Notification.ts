import dayjs, { Dayjs } from 'dayjs'

import Post from '~/models/Post'

export default class Notification {
  id: number
  postId: number
  emoji: string
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  post: Post|null = null

  constructor (data: any) {
    this.id = data.id
    this.postId = data.postId
    this.emoji = data.emoji
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }

    if (data.post) {
      this.post = new Post(data.post)
    }
  }
}
