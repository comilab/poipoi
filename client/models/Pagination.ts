export default class Pagination {
  currentPage: number
  from: number
  lastPage: number
  path: string
  perPage: number
  to: number
  total: number

  constructor (data: any) {
    this.currentPage = data.currentPage
    this.from = data.from
    this.lastPage = data.lastPage
    this.path = data.path
    this.perPage = data.perPage
    this.to = data.to
    this.total = data.total
  }
}
