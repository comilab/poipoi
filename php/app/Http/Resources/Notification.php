<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class Notification extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function toArray($request)
    {
        return [
            'id' => $this->id,
            'postId' => $this->post_id,
            'emoji' => $this->emoji,
            'createdAt' => $this->created_at,
            'updatedAt' => $this->updated_at,

            'post' => new Post($this->whenLoaded('post')),
        ];
    }
}
