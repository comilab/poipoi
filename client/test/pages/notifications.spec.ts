import { render, waitFor } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { ref } from '@vue/composition-api'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import dayjs from 'dayjs'

import Notifications from '~/pages/notifications.vue'
import useNotificationsApi from '~/composables/use-notifications-api'
import postFactory from '~/test/factories/post'
import paginationFactory from '~/test/factories/pagination'
import notificationFactory from '~/test/factories/notification'
import Pagination from '~/models/Pagination'
import Notification from '~/models/Notification'
import useMockedComponent from '~/test/stubs/MockedComponent'

jest.mock('~/composables/use-notifications-api')

describe('pages/notifications', () => {
  let options: any
  let contextMock: any
  let fetchMock: any
  let notificationsApiMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      },
      mocks: {
        $fetchState: {
          pending: false
        }
      }
    }

    contextMock = {
      redirect: jest.fn(),
      route: ref<any>({
        query: {}
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useContext').mockReturnValue(contextMock)

    fetchMock = {
      fetch: jest.fn()
    }
    jest.spyOn(NuxtCompositionApi, 'useFetch').mockReturnValue(fetchMock)

    notificationsApiMock = {
      index: jest.fn(),
      notifications: ref<Notification[]>([]),
      pagination: ref<Pagination|null>(null)
    }
    mocked(useNotificationsApi).mockReturnValue(notificationsApiMock)
  })

  describe('renders correctly', () => {
    test('投稿がある時', () => {
      notificationsApiMock.notifications.value = notificationFactory.buildList(10, {
        emoji: '🍣',
        post: postFactory.build({
          path: '/1',
          caption: 'caption',
          images: []
        }),
        createdAt: dayjs('2020-01-01 00:00')
      })
      notificationsApiMock.pagination.value = paginationFactory.build({
        currentPage: 1,
        perPage: 10,
        total: 100
      })
      options.stubs.AppPagination = true
      const { container } = render(Notifications, options)

      expect(container).toMatchSnapshot()
    })

    test('投稿がない時', () => {
      notificationsApiMock.notifications.value = []
      notificationsApiMock.pagination.value = paginationFactory.build({
        currentPage: 1,
        perPage: 10,
        total: 100
      })
      options.stubs.AppPagination = true
      const { container } = render(Notifications, options)

      expect(container).toMatchSnapshot()
    })
  })

  describe('fetch', () => {
    test('indexが実行される', async () => {
      render(Notifications, options)

      expect(NuxtCompositionApi.useFetch).toBeCalled()

      const callback = mocked(NuxtCompositionApi.useFetch).mock.calls[0][0] as Function
      await callback()

      expect(notificationsApiMock.index).toBeCalled()
    })

    test('route.queryが変更されたら再実行', async () => {
      render(Notifications, options)

      contextMock.route.value = {
        query: {
          page: '1'
        }
      }

      await waitFor(() => {
        expect(fetchMock.fetch).toBeCalled()
      })
    })
  })

  describe('pagination', () => {
    test('selectedされたらリダイレクトする', async () => {
      notificationsApiMock.notifications.value = notificationFactory.buildList(1)
      const { component, emitsList } = useMockedComponent({
        emits: ['selected']
      })
      notificationsApiMock.pagination.value = paginationFactory.build()
      options.stubs.AppPagination = component

      render(Notifications, options)
      emitsList[0].selected(1)

      await waitFor(() => {
        expect(contextMock.redirect).toBeCalledWith('/notifications', {
          page: '1'
        })
      })
    })
  })
})
