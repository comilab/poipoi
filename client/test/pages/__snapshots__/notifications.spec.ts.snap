// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/notifications renders correctly ÊäïÁ®ø„Åå„ÅÇ„ÇãÊôÇ 1`] = `
<div>
  <div
    post="[object Object]"
  >
    <section
      class="rounded bg-white my-8 p-4 max-w-screen-md divide-y mx-auto px-2 md_px-4"
      data-testid="section"
    >
      <!---->
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
      <div
        class="pt-2 mb-2"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded"
          data-testid="button"
          to="/posts/2"
        >
          <!---->
          <div
            class="w-full flex"
            data-testid="text"
          >
            <div
              class="emoji-wrapper inline-block flex-shrink-0 w-8 h-8"
            >
              <img
                alt="üç£"
                class="emoji"
                draggable="false"
                src="https://twemoji.maxcdn.com/v/13.0.1/72x72/1f363.png"
              />
            </div>
            <div
              class="flex-auto overflow-hidden ml-2 md_ml-4"
            >
              <div
                class="mb-1"
              >
                <div
                  class="block md_inline"
                >
                  „É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå„ÅÇ„Çä„Åæ„Åó„Åü
                </div>
                <div
                  class="block md_inline text-sm md_ml-2"
                >
                  (2020-01-01 00:00)
                </div>
              </div>
              <div
                class="text-sm truncate"
              >
                caption
              </div>
            </div>
            <!---->
          </div>
        </nuxtlink-stub>
      </div>
    </section>
    <apppagination-stub
      class="justify-center my-4"
      pagination="[object Object]"
    />
  </div>
</div>
`;

exports[`pages/notifications renders correctly ÊäïÁ®ø„Åå„Å™„ÅÑÊôÇ 1`] = `
<div>
  <div
    post="[object Object]"
  >
    <section
      class="rounded bg-white my-8 p-4 max-w-screen-md divide-y mx-auto px-2 md_px-4"
      data-testid="section"
    >
      <!---->
      <div>
        „ÅäÁü•„Çâ„Åõ„Åå„ÅÇ„Çä„Åæ„Åõ„Çì
      </div>
    </section>
    <apppagination-stub
      class="justify-center my-4"
      pagination="[object Object]"
    />
  </div>
</div>
`;
